buildpack/telemetry/dynatrace.py

```diff
@@ -5,23 +5,148 @@
+    "DT_CLUSTER_ID": None,  # optional, default not set
+    "DT_TAGS": None,  # optional tags e.g. MikesStuff easyTravel=Mike
+    "DT_CUSTOM_PROP": None,  # optional metadata e.g. Department=Acceptance Stage=Sprint
+    # -- Environment variables for troubleshooting
+    "DT_LOGSTREAM": "stdout",  # optional
+    "DT_LOGLEVELCON": None,  # Use this environment variable to define the console log level. Valid options are: NONE, SEVERE, and INFO.
+    "DT_AGENTACTIVE": None,  # Set to true or false to enable or disable OneAgent.
+    # -- Networking environment variables
+    "DT_NETWORK_ZONE": None,  # optional, Specifies to use a network zone. For more information
+    "DT_PROXY": None, # Optional, When using a proxy, use this environment variable to pass proxy credentials.
+}
+
+
+def stage(buildpack_dir, build_path, cache_path):
+    """
+    Downloads and unzips necessary OneAgent components
+    """
+    if is_agent_enabled():
+        try:
+            util.resolve_dependency(
+                "dynatrace.agent",
+                build_path,  # DOT_LOCAL_LOCATION,
+                buildpack_dir=buildpack_dir,
+                cache_dir=cache_path,  # CACHE_DIR,
+                unpack=True,
+                overrides={
+                    "url": os.environ.get("DT_SAAS_URL"),
+                    "environment": os.environ.get("DT_TENANT"),
+                    "token": os.environ.get("DT_PAAS_TOKEN"),
+                },
+                ignore_cache=True
+            )
+        except Exception as e:
+            logging.warning(
+                "Dynatrace agent download and unpack failed", exc_info=True
+            )
+
+
+def update_config(m2ee, app_name):
+    """
+    Injects Dynatrace configuration to java runtime
+    """
+    if not is_agent_enabled():
+        logging.debug(
+            "Skipping Dynatrace OneAgent setup, required env vars are not set"
+        )
+        return
+    logging.info("Enabling Dynatrace OneAgent")
+    try:
+        manifest = get_manifest()
+    except Exception as e:
+        logging.warning(
+            "Failed to parse Dynatrace manifest file", exc_info=True
+        )
+        return
+
+    # dynamic default
+    default_env.update({"DT_TENANTTOKEN": manifest.get("tenantToken")})
+
+    for key, dv in default_env.items():
+        value = os.environ.get(key, dv)
+        if value:
+            util.upsert_custom_environment_variable(m2ee, key, value)
+    util.upsert_custom_environment_variable(
+        m2ee, "DT_CONNECTION_POINT", get_connection_endpoint()
+    )
+
+    agent_path = os.path.join(".local", get_agent_path())
+    if not os.path.exists(agent_path):
+        raise Exception(
+            "Dynatrace Agent not found: {agent_path}".format(agent_path)
+        )
+
+    util.upsert_javaopts(
+        m2ee,
+        [
+            "-agentpath:{path}".format(path=os.path.abspath(agent_path)),
+            "-Xshare:off",
+        ],
+    )
+
+
+def get_manifest():
```

Comment:
- added cache with 3 as maxsize, just not to think about invalidation, probably not needed but no harm.
- updated path to be based on installation path of dynatrace
- not in favor of adding key-value fetch here in this method since the json has nested dictionaries and every user of this method takes care of their need

Reply:
`functools.lru_cache` is a good approach, but let's use the default max size, else when we add new calls it will have the same issue.

Just so you know `lru_cache` doesn't cache if there is an exception, so all the calls would try to open the file and fail. So won't help in those scenarios.

Reply:
changed